[ziyi22] iP

ziyi22 opened the pull request:
# **Judy**
> Hey! I'm Judy, your...

## **Why Judy?**
- efficient
- cute
- ~~EASY~~ SUPER EASY to use
- and **Free** !

## **How to use?**
1. Download Judy here.
2. Navigate to your Downloads folder.
3. Type `java -jar Judy.jar` in your terminal.
4. Click Enter.
5. Add your tasks.
6. let it manage your tasks for you 😉

## **Judy's Features**
- [x] Add tasks (Todo, Event and Deadline)
- [x] Delete task
- [x] List all tasks
- [x] Mark task or Unmark task
- [x] Find a task using keyword
- [ ] Reminder (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ziyi22 committed:
Level-0: Rename, Greet, Exit

ziyi22 committed:
Level-1: Echo

ziyi22 committed:
Level-2: Add, List

ziyi22 committed:
Level-3: Mark as done

ziyi22 committed:
Level-4: ToDos, Events, Deadlines

ziyi22 committed:
A-TextUiTesting: Automated Text UI Testing

ziyi22 committed:
Level-5: Handle Errors

ziyi22 committed:
Level-6: Delete

ziyi22 committed:
Fix bugs

ziyi22 committed:
Add level 7: load tasks and save tasks

camille-readbean commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+            System.out.println("____________________________________");
+            taskManager.saveTask();
+
+        } while (true);
```
This can written as a _switch case_  for better readability, example:

```java
String input = scanner.nextLine().split(" ");
command = input.split(" ")[0];
switch (command.toLowerCase()) {
case "bye":
   ...
case ...:
   ...
default:
   ...
}
```

camille-readbean commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+import java.io.Serializable;
+
+public class Task implements Serializable {
```
Interesting use of `Serializable` interface!  👍

camille-readbean commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+            tasks = (ArrayList<Task>) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            // Handle file not found or corrupted data file
+            //System.err.println("Error loading tasks: " + e.getMessage());
```
Would be good to remove deadcode. However I think in this case you might actually want it to be there to tell the user what happened!  
  
https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#delete-dead-code

camille-readbean commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+
+    public void markTask(String command) throws DukeException {
+        String[] words = command.split(" ");
+        if (words.length == 2) {
+            int taskId = Integer.parseInt(words[1]) -1;
+            if(taskId < tasks.size()) {
+                Task t = tasks.get(taskId);
+                t.markAsDone();
+                System.out.println(" Nice ! I've marked this task as done: \n" + t);
+            } else {
+                throw new DukeException(" Invalid task index. Type 'list' to list out your tasks. ");
+            }
+        } else {
+            throw new DukeException(" Invalid format. Type 'mark 1' to mark task 1. ");
+        }
```
Might want to avoid deep nesting 😄 

By using guard clauses you can do stuff like:

```java
if (words.length != 2) {
    throw new ...
}
// rest of your code
...
```

The deeper the nesting, the harder it is for the keep track of the logic!

Deep nesting: ℹ️ 
https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting

camille-readbean commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+                t.markAsUndone();
+                System.out.println(" Ok, I've marked this task as not done yet: \n" + t);
+            } else {
+                throw new DukeException(" Invalid task index. Type 'list' to list out your tasks. ");
```
I realised these `DukeException` are the same in `unmarkTask` and `markTask` you can make a new subclass that inherits `DukeException` with the error message declared in the class. 
  
So if you ever need to change the message you or reuse it you can just use the class.

camille-readbean commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+
+    public void addTask(String command) throws DukeException {
+        Task t;
+        switch (getTaskType(command)) {
+            case "todo":
+                t = handleTodo(command);
+                break;
+            case "deadline":
+                t = handleDeadline(command);
+                break;
+            case "event":
+                t = handleEvent(command);
+                break;
+            default:
+                throw new DukeException(" Sorry! I don't know what that means :c ");
+        }
```
Might want to reduce the indentation of case.  
  
```
switch (cond) {
case xxx:
     ...
case yyyy:
    ...
default:
    ....
}
```  

See:  
https://se-education.org/guides/conventions/java/intermediate.html#layout

camille-readbean commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,56 @@
+        Scanner scanner = new Scanner(System.in);
+        String command;
+
+        do {
+            try {
+                command = scanner.nextLine();
+                System.out.println("____________________________________");
+                if(command.equalsIgnoreCase("bye")) {
+                    break;
+                } else if (command.equals("list")) {
+                    taskManager.listOutTasks();
+                } else if (command.startsWith("mark")){
+                    taskManager.markTask(command);
+                } else if (command.startsWith("unmark")) {
+                    taskManager.unmarkTask(command);
+                } else if (command.startsWith("delete")) {
+                    taskManager.deleteTask(command);
+                } else {
+                    taskManager.addTask(command);
+                }
+            } catch (DukeException e) {
+                System.out.println(e.getMessage());
+            }
```
This can written as a _switch case_  for better readability, example:

```java
String input = scanner.nextLine().split(" ");
command = input.split(" ")[0];
switch (command.toLowerCase()) {
case "bye":
   ...
case ...:
   ...
default:
   ...
}
```

see:  
https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting

camille-readbean commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    public void listOutTasks() {
+        if (tasks.isEmpty()) {
+            System.out.println(" No task added yet!");
+        } else {
+            System.out.println(" Here are the tasks in your list: ");
+            for(int i = 0; i < tasks.size(); i++) {
+                int seq = i+1;
+                Task t = tasks.get(i);
+                System.out.println(" " + seq + ". " + t.toString());
+            }
+        }
```
Coding quality, make the happy path prominent. 

No need for the `else` for `!command.contains(" ")`

You can do like a guard clause style for easier readability and avoid complicated nesting.  
  
```java
if (tasks.isEmpty()) {
    // fail check
    System.out.println();
    return;
}
// rest of your code
...
```

The `return` will ensure the remaining code will not be executed anyways so no need for else. 

https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent

camille-readbean reviewed (commented):
Code seems good.  👍   
Keep it up! 😄   
  
Some points to take note of, if-else are good but don't nest too much of them.  
  
Take note i think your taskList.txt is **not actually plaintext**.  
So it might be wrong to actually give it as `.txt` file extension.  
This is because serialisation stores the Java object.  

https://nus-cs2103-ay2324s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting

ziyi22 committed:
Add level 8:Date Time Format

ziyi22 committed:
Add level 8:Date Time Format

RyanNgWH commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,25 @@
+import java.time.format.DateTimeParseException;
+
+public class Deadline extends Task {
+    protected LocalDate by;
```
Perhaps use a more descriptive variable name such as `dueDate`?

RyanNgWH commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    public void deleteTask(String command) throws DukeException {
+        String[] parts = command.trim().split(" ");
+        if (parts.length == 2) {
+            int taskIndex = Integer.parseInt(parts[1].trim()) - 1;
+            if(taskIndex < tasks.size()) {
+                Task removedTask = tasks.remove(taskIndex);
+                System.out.println(" Noted. I've removed this task:\n" + removedTask);
+                printTaskSize();
+            } else {
+                throw new DukeException(" Invalid task index");
+            }
+        } else {
+            throw new DukeException(" Invalid delete format. Try: delete <Task Index> ");
+        }
+    }
+    private String getTaskType(String command) {
+        return command.trim().split("\\s+")[0].toLowerCase();
+    }
```
Could consider adding line breaks between statements to improve readability

```java
    public void deleteTask(String command) throws DukeException {
        String[] parts = command.trim().split(" ");

        if (parts.length == 2) {
            int taskIndex = Integer.parseInt(parts[1].trim()) - 1;

            if(taskIndex < tasks.size()) {
                Task removedTask = tasks.remove(taskIndex);

                System.out.println(" Noted. I've removed this task:\n" + removedTask);
                printTaskSize();
            } else {
                throw new DukeException(" Invalid task index");
            }
        } else {
            throw new DukeException(" Invalid delete format. Try: delete <Task Index> ");
        }
    }

    private String getTaskType(String command) {
        return command.trim().split("\\s+")[0].toLowerCase();
    }
```

RyanNgWH commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,57 @@
+        System.out.println("________________________________________");
+    }
+
+    private static void echo() {
```
Maybe rename this method to better reflect the functionality since it is no longer echoing the user input

RyanNgWH commented on src/main/java/Tasks.java:
```diff
@@ -0,0 +1,173 @@
+
+public class Tasks implements Serializable {
+    private static ArrayList<Task> tasks;
+    private static final String FILE_PATH = "src/data/taskList.txt";
```
Good use of SCREAMING_SNAKE_CASE for constant variable!

RyanNgWH commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,5 @@
+public class DukeException extends Exception{
```
Missing space between `Exception` and `{`

Noticed a few instances of this elsewhere as well

RyanNgWH reviewed (commented):
Code is overall well formatted and following the coding standards. Good job!!!

Just a few instances of missed spaces and adding line breaks could greatly improve code readability.

ziyi22 committed:
Add A-MoreOOP: Reorganize code and extract more classes

ziyi22 committed:
Merge branch 'A-MoreOOP'

ziyi22 committed:
Add Help Command

ziyi22 committed:
Organize classes into packages: Add Decoder and Encoder classes

ziyi22 committed:
Merge branch 'branch-A-Packages'

ziyi22 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ziyi22 committed:
Set up Gradle and run Duke using Gradle

ziyi22 committed:
Add JUnit Tests: Test the parsing of and the executing of commands

ziyi22 committed:
Add JUnit Tests: Test the parsing of and the executing of commands

ziyi22 committed:
Merge branch 'branch-A-JUnit'

ziyi22 committed:
Fix filepath not found error

ziyi22 committed:
Merge branch 'branch-A-Jar'

ziyi22 committed:
Add JavaDoc comments.

ziyi22 committed:
Merge branch 'branch-A-JavaDoc'

ziyi22 committed:
Improve code to follow the coding standard

ziyi22 committed:
Merge branch 'branch-A-CodingStandard'

ziyi22 committed:
Implement Level-9: Find task by searching for a keyword

ziyi22 committed:
Merge branch 'branch-Level-9'

ziyi22 committed:
Implements A-CheckStyle: solve coding style violations

ziyi22 committed:
Implement A-CheckStyle: solve coding style violations

ziyi22 committed:
Merge branch 'branch-A-CheckStyle'

ziyi22 committed:
Implement Level-10: GUI

ziyi22 committed:
Implement Level 10: GUI and JavaFX.

ziyi22 committed:
Merge branch 'branch-Level-10'

ziyi22 committed:
Add assertions to enhance code reliability in the Parser class

Introduce assertions in the Parser class to validate assumptions about the input parameters and internal state during the parsing process. Specifically, assertions have been added to ensure:

1. Valid mark, unmark, and delete command formats with the correct number of command parts.
2. Validity of task indices to prevent ArrayIndexOutOfBoundsException and NumberFormatException.
3. Proper format for todo, deadline, and event commands to avoid unexpected errors.
4. Consistency in the structure of input commands to facilitate correct parsing.

These assertions aim to improve code robustness and aid in identifying potential issues during development.

ziyi22 committed:
Merge pull request #2 from ziyi22/branch-A-Assertions

Add assertions to enhance code reliability in the Parser class

ziyi22 committed:
Improve code quality: reduce nesting, simplify expressions, and enhance exception handling

Refactor the Parser class to enhance code readability, maintainability, and error handling. The following changes have been made:

1. **Reduced Nesting:** Reorganized code to reduce nested levels, making the logic more straightforward and easier to understand.

2. **Simplified Expressions:** Streamlined complex expressions and conditions within the parse method, improving code conciseness and comprehensibility.

3. **Exception Handling:** Introduced additional exception classes to handle specific error scenarios more gracefully. Exceptions are now thrown with informative error messages, providing better feedback to users and aiding in debugging. The newly added exception classes include:
   - `InvalidTaskIndexException`: For cases where an invalid task index is provided.
   - `InvalidTodoException`: For cases where a todo command has an empty description.
   - `InvalidDeadlineException`: For cases where a deadline command has an empty description or an invalid format.
   - `InvalidEventException`: For cases where an event command has an empty description or an invalid format.
   - `InvalidDateTimeException`: For cases where parsing a date-time encounters an invalid format.

4. **Added parseDateTime Method:** Introduced a new method `parseDateTime` to parse date-time strings into `LocalDateTime` objects, handling potential exceptions and throwing a custom `InvalidDateTimeException` for invalid formats.

5. **Introduced isEmptyDescription Method:** Added a method `isEmptyDescription` to check if user input has an empty description, simplifying validation conditions.

These changes collectively aim to improve code quality, enhance maintainability, and provide better feedback to users in case of errors.

ziyi22 committed:
Merge pull request #3 from ziyi22/branch-A-CodeQuality

Improve code quality: reduce nesting, simplify expressions, and enhan…

ziyi22 committed:
Improve code quality: reduce nesting, simplify expressions, and enhance exception handling

Refactor the Parser class to enhance code readability, maintainability, and error handling. The following changes have been made:

1. **Reduced Nesting:** Reorganized code to reduce nested levels, making the logic more straightforward and easier to understand.

2. **Simplified Expressions:** Streamlined complex expressions and conditions within the parse method, improving code conciseness and comprehensibility.

3. **Exception Handling:** Introduced additional exception classes to handle specific error scenarios more gracefully. Exceptions are now thrown with informative error messages, providing better feedback to users and aiding in debugging. The newly added exception classes include:
   - `InvalidTaskIndexException`: For cases where an invalid task index is provided.
   - `InvalidTodoException`: For cases where a todo command has an empty description.
   - `InvalidDeadlineException`: For cases where a deadline command has an empty description or an invalid format.
   - `InvalidEventException`: For cases where an event command has an empty description or an invalid format.
   - `InvalidDateTimeException`: For cases where parsing a date-time encounters an invalid format.

4. **Added parseDateTime Method:** Introduced a new method `parseDateTime` to parse date-time strings into `LocalDateTime` objects, handling potential exceptions and throwing a custom `InvalidDateTimeException` for invalid formats.

5. **Introduced isEmptyDescription Method:** Added a method `isEmptyDescription` to check if user input has an empty description, simplifying validation conditions.

These changes collectively aim to improve code quality, enhance maintainability, and provide better feedback to users in case of errors.

ziyi22 committed:
Improve code quality: reduce nesting, simplify expressions, and enhance exception handling

Refactor the Parser class to enhance code readability, maintainability, and error handling. The following changes have been made:

1. **Reduced Nesting:** Reorganized code to reduce nested levels, making the logic more straightforward and easier to understand.

2. **Simplified Expressions:** Streamlined complex expressions and conditions within the parse method, improving code conciseness and comprehensibility.

3. **Exception Handling:** Introduced additional exception classes to handle specific error scenarios more gracefully. Exceptions are now thrown with informative error messages, providing better feedback to users and aiding in debugging. The newly added exception classes include:
   - `InvalidTaskIndexException`: For cases where an invalid task index is provided.
   - `InvalidTodoException`: For cases where a todo command has an empty description.
   - `InvalidDeadlineException`: For cases where a deadline command has an empty description or an invalid format.
   - `InvalidEventException`: For cases where an event command has an empty description or an invalid format.
   - `InvalidDateTimeException`: For cases where parsing a date-time encounters an invalid format.

4. **Added parseDateTime Method:** Introduced a new method `parseDateTime` to parse date-time strings into `LocalDateTime` objects, handling potential exceptions and throwing a custom `InvalidDateTimeException` for invalid formats.

5. **Introduced isEmptyDescription Method:** Added a method `isEmptyDescription` to check if user input has an empty description, simplifying validation conditions.

These changes collectively aim to improve code quality, enhance maintainability, and provide better feedback to users in case of errors.

ziyi22 committed:
Add User Guide

ziyi22 committed:
Merge pull request #4 from ziyi22/branch-A-UserGuide

Implement A-User Guide

ziyi22 committed:
Edit User Guide

ziyi22 committed:
Update README.md

ziyi22 committed:
Update README.md

ziyi22 committed:
Add Ui.png

ziyi22 committed:
Merge remote-tracking branch 'origin/master'

ziyi22 committed:
Edit README.md
